config/src/main/java/com/typesafe/config/impl/BeanFactory.java

```diff
@@ -0,0 +1,160 @@
+            originalNames.put(toCamelCase(configProp.getKey()),configProp.getKey());
+        }
+
+        try {
+            T bean = clazz.newInstance();
+            for (PropertyDescriptor beanProp : Introspector.getBeanInfo(clazz).getPropertyDescriptors()) {
+                if (beanProp.getReadMethod() == null || beanProp.getWriteMethod() == null) {
+                    continue;
+                }
+                Method setter = beanProp.getWriteMethod();
+                Object configValue = configProps.get(beanProp.getName());
+                if (configValue == null) {
+                    throw new ConfigException.Generic(
+                            "Could not find " + beanProp.getName() + " from " + clazz.getName() + " in config.");
+                }
+                if (configValue instanceof Map) {
+                    configValue = createInternal(config.getConfig(originalNames.get(beanProp.getDisplayName())), beanProp.getPropertyType());
+                } else {
+                    Class parameterClass = setter.getParameterTypes()[0];
+                    configValue = getValueWithAutoConversion(parameterClass, config, originalNames.get(beanProp.getDisplayName()));
+                }
+                setter.invoke(bean, configValue);
+
+            }
+            return bean;
+        } catch (IntrospectionException e) {
+            throw new ConfigException.Generic("Could not resolve a string method name.", e);
+        } catch (InstantiationException e) {
+            throw new ConfigException.Generic(clazz + " needs a public no args constructor", e);
+        } catch (IllegalAccessException e) {
+            throw new ConfigException.Generic(clazz + " getters and setters are not accessible", e);
+        } catch (InvocationTargetException e) {
+            throw new ConfigException.Generic("Calling bean method caused exception", e);
+        }
+    }
+
+    private static Object getValueWithAutoConversion(Class parameterClass, Config config, String configPropName) {
+        if (parameterClass == Boolean.class || parameterClass == boolean.class) {
+            return config.getBoolean(configPropName);
+        } else if (parameterClass == Integer.class || parameterClass == int.class) {
+            return config.getInt(configPropName);
+        } else if (parameterClass == Double.class || parameterClass == double.class) {
+            return config.getDouble(configPropName);
+        } else if (parameterClass == Long.class || parameterClass == long.class) {
+            String rawVal = config.getString(configPropName);
+            if (isDuration(rawVal)) {
+                TimeUnit timeUnit = getTimeUnit(rawVal);
+                return config.getDuration(configPropName, timeUnit);
+            } else if (isByteValue(rawVal)) {
+                return config.getBytes(configPropName);
+            }
+            return config.getLong(configPropName);
```

Comment:
really I think the best way to do this would to have it be a java.time.Duration instead of a long. the real problem here is that we're overloading long and using it not a in typesafe manner. of course to use java.time.Duration we'd have to support Java 8. i think it wouldn't be a bad call to support Java 8 because Java 7 is EOL (end of life) in April. Play is currently discussing releasing Play 2.4 only for Java 8 for that reason. See https://groups.google.com/forum/#!topic/play-framework-dev/fte3TC4rYNw